opal/mca/btl/iwarp/btl_iwarp_component.c

```diff
@@ -0,0 +1,3788 @@
+            mtts_per_seg = 1ull << read_module_param("/sys/module/mlx4_core/parameters/log_mtts_per_seg", 1, 63);
+            num_mtt = 1ull << read_module_param("/sys/module/mlx4_core/parameters/log_mtts_per_seg", 1, 63);
+            if (1 == num_mtt) {
+                /* NTH: is 19 a minimum? when log_num_mtt is set to 0 use 19 */
+                num_mtt = 1 << 19;
+                max_reg = (num_mtt - reserved_mtt) * opal_getpagesize () * mtts_per_seg;
+            } else  {
+                max_reg = (num_mtt - reserved_mtt) * opal_getpagesize () * mtts_per_seg;
+            }
+        }
+
+    } else if (!strncmp(device_name, "mthca", 5)) {
+        if (0 == stat("/sys/module/ib_mthca/parameters/num_mtt", &statinfo)) {
+            mtts_per_seg = 1ull << read_module_param("/sys/module/ib_mthca/parameters/log_mtts_per_seg", 1, 63);
+            num_mtt = read_module_param("/sys/module/ib_mthca/parameters/num_mtt", 1 << 20, (uint64_t) -1);
+            reserved_mtt = read_module_param("/sys/module/ib_mthca/parameters/fmr_reserved_mtts", 0, (uint64_t) -1);
+
+            max_reg = (num_mtt - reserved_mtt) * opal_getpagesize () * mtts_per_seg;
+        } else {
+            max_reg = mem_total;
+        }
+
+    } else {
+        /* Need to update to determine the registration limit for this
+           configuration */
+        max_reg = mem_total;
+    }
+
+    /* Print a warning if we can't register more than 75% of physical
+       memory.  Abort if the abort_not_enough_reg_mem MCA param was
+       set. */
+    if (max_reg < mem_total * 3 / 4) {
+        char *action;
+
+        if (mca_btl_iwarp_component.abort_not_enough_reg_mem) {
+            action = "Your MPI job will now abort.";
+        } else {
+            action = "Your MPI job will continue, but may be behave poorly and/or hang.";
+        }
+        opal_show_help("help-mpi-btl-iwarp.txt", "reg mem limit low", true,
+                       opal_process_info.nodename, (unsigned long)(max_reg >> 20),
+                       (unsigned long)(mem_total >> 20), action);
+        return 0;  /* signal that we can't have enough memory */
+    }
+
+    /* Limit us to 87.5% of the registered memory (some fluff for QPs,
+       file systems, etc) */
+    return (max_reg * 7) >> 3;
+}
+
+static int init_one_device(opal_list_t *btl_list, struct ibv_device* ib_dev)
+{
+    mca_rcache_base_resources_t rcache_resources;
+    mca_btl_iwarp_device_t *device;
+    uint8_t i, k = 0;
+    int ret = -1, port_cnt;
+    opal_btl_iwarp_ini_values_t values, default_values;
+    int *allowed_ports = NULL;
+    bool need_search;
+    struct ibv_context *dev_context = NULL;
+
+    /* Open up the device */
+    dev_context = ibv_open_device(ib_dev);
+    if (NULL == dev_context) {
+        return OPAL_ERR_NOT_SUPPORTED;
+    }
+
+    /* Find out if this device supports RC QPs */
+    if (OPAL_SUCCESS != opal_common_verbs_qp_test(dev_context,
```

Comment:
When would we ever open an iWARP device where RC QPs were not supported?  This feels like dead code (I think this check was added to the openib BTL so that it would skip usNIC devices -- because usNIC does not support RC).